### ISSUE TITLE
Mustache templates for tables using script[type=text/plain] are corrupted

### ISSUE THREAD

**Issue by westonruter** [ampproject/amp-wp#4254]
## Bug Description

As reported on a [support forum topic](https://wordpress.org/support/topic/in-mustache-templates-are-removed-all-closing-tags/) by @spasibych, a workaround for Mustache templates for tables is broken by the DOM parser in the AMP plugin. 

Please refer to the [amp.dev docs on Mustache templates for tables](https://amp.dev/documentation/components/amp-mustache/#tables).

In particular, given this input:

```html
<script type="text/plain" template="amp-mustache">
  <table>
    <tr>
{{#foo}}<td></td>{{/foo}}
    </tr>
  </table>
</script>
```

The AMP plugin is outputting this as:

```html
<script type="text/plain" template="amp-mustache">
  <table>
    <tr>
{{#foo}}<td>{{/foo}}		
</script>
```

However, the use of a `script` to contain the template is _one_ of the workarounds mentioned:

> Workarounds include wrapping Mustache sections in HTML comments (e.g. `<!-- {{#bar}} -->`), using non-table elements like `<div>` instead or using a `<script type="text/plain">` tag to define your templates.

It turns out that that the HTML comment workaround does work. Given this input:

```html
<template type="amp-mustache">
  <table>
    <tr>
<!--{{#foo}}--><td></td><!--{{/foo}}-->
    </tr>
  </table>
</template>
```

The AMP plugin makes no modifications in the output.

## Expected Behaviour

The mustache template for a table using the `<script type="text/plain">` workaround should work.

## Steps to reproduce

1. Create a Custom HTML block with the above `<script type="text/plain" template="amp-mustache">...</script>` element.
2. View the AMP page.
3. Notice that the closing tags are dropped.

## Additional context

 - WordPress version:
 - Plugin version:
 - Gutenberg plugin version (if applicable):
 - AMP plugin template mode:
 - PHP version:
 - OS:
 - Browser: [e.g. chrome, safari]

---------------

_Do not alter or remove anything below. The following sections will be managed by moderators only._

## Acceptance criteria

* 

## Implementation brief

As noted in https://github.com/ampproject/amphtml/issues/26656#issuecomment-584373338:

We'll have to figure out some other mechanism to workaround libxml failing to parse the HTML. Something that comes to mind is doing a search/replace to convert `<script type="text/plain" template="amp-mustache">` into some other element (e.g. `tmp-script`) during parsing, and then replace that placeholder with the original `script` once the DOM has been constructed.

This turns out to work: https://3v4l.org/0qq5S

```php
<?php

$script = '
    <script type="text/plain" template="amp-mustache">
      <table>
        <tr>
            {{#foo}}<td></td>{{/foo}}
        </tr>
      </table>
    </script>
';

$dom = new DOMDocument();

$tmp_script = preg_replace( 
    '#<script(\s[^>]*template="amp-mustache"[^>]*)>(.*?)</script>#s', 
    '<tmp-script$1>$2</tmp-script>',
    $script
);

@$dom->loadHTML( '<!DOCTYPE html><html><head><meta charset="utf-8"></head><body>' . $tmp_script . '</body></html>' );

$tmp_script_elements = iterator_to_array( $dom->getElementsByTagName( 'tmp-script' ) );
foreach ( $tmp_script_elements as $tmp_script_element ) {
    $script = $dom->createElement( 'script' );
    foreach ( $tmp_script_element->attributes as $attr ) {
        $script->setAttribute( $attr->nodeName, $attr->nodeValue );
    }
    while ( $tmp_script_element->firstChild ) {
        $script->appendChild( $tmp_script_element->firstChild );
    }
    $tmp_script_element->parentNode->replaceChild( $script, $tmp_script_element );
}

echo $dom->saveHTML();
```

Note the regex for doing the search/replace will need to be hardened, in particular the regex for matching the attributes (in any order).

The logic should go into the `Document` class and called similarly to the `maybe_replace_noscript_elements` method:

https://github.com/ampproject/amp-wp/blob/ea2814819225b375cdb97838538fde8fe3f94de9/src/Dom/Document.php#L602-L639

Note that this replacement must be done immediately after parsing, since the `Document` will then need to have the `script` replaced to successfully go through the sanitizers.

## QA testing instructions

* 

## Demo

* 

## Changelog entry

*

**Comment by westonruter** [comment]
> However, the use of a `script` to contain the template is _one_ of the workarounds mentioned:
> 
> > Workarounds include wrapping Mustache sections in HTML comments (e.g. `<!-- {{#bar}} -->`), using non-table elements like `<div>` instead or using a `<script type="text/plain">` tag to define your templates.
> 
> It turns out that that the HTML comment workaround does work.

Actually, it only works on origin. The Google AMP Cache and the AMP Optimizer both strip out the HTML comments. See https://github.com/ampproject/amphtml/issues/26656.

**Comment by kienstra** [comment]
Weston's [Steps To Reproduce](https://github.com/ampproject/amp-wp/issues/4254#issue-561789003) should be good for testing instructions.

**Comment by csossi** [comment]
Verified in QA:
![image](https://user-images.githubusercontent.com/16945664/76714287-9e30a000-66fc-11ea-8e5d-db14ecac858a.png)

**Comment by spasibych** [comment]
Bug is present with plugin version 1.5 (1.5.0 and 1.5.1).

**Comment by westonruter** [comment]
Bizzare. We did fix this. But I'm not seeing it now.

**Comment by westonruter** [comment]
@spasibych Here’s a 1.5.2-RC1 pre-release build with the fix for you to test: [amp.zip](https://github.com/ampproject/amp-wp/files/4429512/amp.zip) (v1.5.2-RC1-20200403T193031Z-3102c5dea)

**Comment by westonruter** [comment]
We've just published the 1.5.2 release to WordPress.org: https://wordpress.org/plugins/amp/

You can update now via the WordPress admin.

Release notes: https://github.com/ampproject/amp-wp/releases/tag/1.5.2

**Comment by grapheon** [comment]
@westonruter now everything is fine, thank you very much!
